### PR TITLE
MRVA: Get more accurate links using `sourceLocationPrefix`

### PR DESCRIPTION
**Pull request by shati-patel**
The MRVA action includes the `sourceLocationPrefix` in the result index. This PR updates the VS Code interface to use that prefix to construct links (instead of making a "best guess" based on the expected format). This is only used for raw (non-interpreted) results.

Main change is in https://github.com/github/vscode-codeql/commit/7b9e6a8effb57d08c34b4a0d4bb348f39b806ed2 (the other commits just pass `sourceLocationPrefix` around through various functions).

There shouldn't be any visible change, so I've tested this by running some queries and checking the resulting links. I've also updated the test data for markdown generation to use a `sourceLocationPrefix` for one of the repos, and the rendered links look fine there too.

See internal issue for more details.

## Checklist

N/A - internal only 🦒 

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

**Review by charisk — APPROVED** [review]
Perfect PR! Description, code, commits and testing all great!

**Review by aeisenberg — COMMENTED** [review]
Can you add some unit tests specifically for `tryGetRemoteLocation`? There's complex string manipulation and it would be good to have some proof that it does what you expect, especially for edge cases.

File: extensions/ql-vscode/src/pure/bqrs-utils.ts [hunk 1]
```diff
@@ -97,20 +97,29 @@ export function isStringLoc(loc: UrlValue): loc is string {
...
-  // So we need to drop the first 6 parts of the path.
+  let trimmedLocation: string;
 
-  // TODO: We can make this more robust to other path formats.
-  const locationParts = resolvableLocation.uri.split('/');
-  const trimmedLocation = locationParts.slice(6, locationParts.length).join('/');
+  // Remote locations have the following format:
+  // "file:${sourceLocationPrefix}/relative/path/to/file"
+  // So we need to strip off the first part to get the relative path.
+  if (sourceLocationPrefix) {
```
[the remark this answers is not in the document]
Comment (shati-patel) [reply extensions/ql-vscode/src/pure/bqrs-utils.ts:113]: Technically it could be undefined (for older queries that were run before the upstream change, ie before last week), but in that case the `if (...)` will still handle those as expected.

**Comment by shati-patel** [comment]
> Can you add some unit tests specifically for `tryGetRemoteLocation`? There's complex string manipulation and it would be good to have some proof that it does what you expect, especially for edge cases.

Tests incoming! 🏃🏽‍♀️ 

Your comments also made me realise that the previous code was missing appropriate error handling/checks etc 🙈  Not quite part of this PR, but I'll push up some fixes here too.

**Comment by shati-patel** [comment]
Updates! (sorry for bulking out this PR a bit more 🙃)
- https://github.com/github/vscode-codeql/pull/1367/commits/67dd8cb3b144bf78fd4f4129762378a9dad26726 checks that the file URI looks sensible, before we manipulate it.
- https://github.com/github/vscode-codeql/pull/1367/commits/60bc581999f5bfe3af9e1bf4e5f0d3f11f3c4fcd makes sure that we don't try to render a link if the URL is `undefined`.
- https://github.com/github/vscode-codeql/pull/1367/commits/37852f308ff1df2bc35150bef76ef4c3f969dc97 adds tests for `tryGetRemoteLocation`.
  - ❓ Naming question: I was a bit confused by the testing structure 😅 (e.g. I was expecting to add tests in a `bqrs-utils.test.ts` file, but instead found existing tests in `location.test.ts`, so I've added mine there too)
  Would it make sense to either rename `bqrs-utils.ts` to `location.ts`, or `location.test.ts` to `bqrs-utils.test.ts` so that the file names align?

File: extensions/ql-vscode/src/remote-queries/view/RawResultsTable.tsx [hunk 2]
```diff
@@ -48,7 +48,11 @@ const Cell = ({
       return <span>{convertNonPrintableChars(value.toString())}</span>;
     case 'object': {
       const url = tryGetRemoteLocation(value.url, fileLinkPrefix, sourceLocationPrefix);
-      return <Link href={url}>{convertNonPrintableChars(value.label)}</Link>;
+      if (url) {
+        return <Link href={url}>{convertNonPrintableChars(value.label)}</Link>;
+      } else {
+        return <span>{convertNonPrintableChars(value.label)}</span>;
+      }
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/remote-queries/view/RawResultsTable.tsx:55]: minor:

```suggestion
      const safeLabel = convertNonPrintableChars(value.label);
      if (url) {
        return <Link href={url}>{safeLabel}</Link>;
      } else {
        return <span>{safeLabel}</span>;
      }
```

**Review by aeisenberg — COMMENTED** [review]
Two minor suggestions and 🚢 !

Thanks for adding the tests.

**Review by aeisenberg — APPROVED** [review]
To minimize review cycles, I'll approve now. My comment about changing the parameter type is something I'd like to see. The second comment about extracting the local variable is optional.

**Comment by shati-patel** [comment]
> To minimize review cycles, I'll approve now. My comment about changing the parameter type is something I'd like to see. The second comment about extracting the local variable is optional.

Both done, thank you! 🎉 ✨
